brainiak/hyperparamopt/hpo.py

```diff
@@ -0,0 +1,231 @@
+    s = 0
+    while (np.abs(x - samples[ei.argmax()]).min() < h):
+        ei[ei.argmax()] = 0
+        s = s + 1
+        if (s == samples.size):
+            break
+    xnext = samples[ei.argmax()]
+
+    return xnext
+
+
+def getSample(x, y, dist, minlimit=-np.inf, maxlimit=np.inf):
+    if (dist == 'GMM'):
+        return getNextSample(x, y, minlimit, maxlimit)
+    if (dist == 'uniform'):
+        return np.random.random() * (maxlimit - minlimit) + minlimit
+    if (dist == 'loguniform'):
+        return np.exp(np.random.random()
+                      * (np.log(maxlimit) - np.log(minlimit))
+                      + np.log(minlimit))
+
+
+def fmin(lossfn,
+         space,
+         maxevals,
+         trials,
+         init_random_evals=30,
+         explore_prob=0.2,
+         verbose=False):
+    """Find the minimum of function through hyper paramter optimization
+
+    Arguments
+    ---------
+
+    lossfn : function that takes in a dictionary and returns a real value
+             Function to be minimized
+
+    space : Dictionary specifying the range and distribution of
+            the hyperparamters
+
+    maxevals : int
+               Maximum number of evaluations of lossfn allowed
+
+    trials : list
+             Holds the output of the optimization trials
+             Need not be empty to begin with, new trials are appended
+             at the end
+
+    init_random_evals : int, default 30
+                        Number of random trials to initialize the
+                        optimization
+
+    explore_prob : double in [0, 1], default 0.2
+                   Controls the exploration-vs-exploitation ratio
+                   Currently 20% of trails are random samples
+
+    verbose : bool, default False
+              Get information on current point being processed
+
+    Returns
+    -------
+
+    best : trial entry (dictionary of hyperparameters)
+           Best hyperparameter setting found
+    """
+
+    for s in space:
+        if (space[s]['dist'] is not 'uniform' and
+                space[s]['dist'] is not 'loguniform'):
+            logger.error('Unsupported distribution for variable')
+            raise TypeError('Unknown distribution type for variable')
+
+    if (len(trials) > init_random_evals):
+        init_random_evals = 0
+
+    for t in tqdm(range(maxevals)):
+        sdict = {}
+
+        if (t >= init_random_evals and np.random.random() > explore_prob):
+            search_algo = 'Exploit'
+        else:
+            search_algo = 'Explore'
+
+        yarray = np.array([tr['loss'] for tr in trials])
+        for s in space:
+            sarray = np.array([tr[s] for tr in trials])
+            dist = 'GMM' if (search_algo == 'Exploit') else space[s]['dist']
+            sdict[s] = getSample(sarray, yarray, dist,
+                                 minlimit=space[s]['lo'],
+                                 maxlimit=space[s]['hi'])
+
+        if (verbose):
```

Comment:
It would be more normal to skip checking the verbose flag and just log at debug level rather than info.